proposals/0000-allow-constraint-type-synonyms-in-deriving-heads.md

```diff
@@ -0,0 +1,200 @@
+---
+author: Grigorii Gerasev
+date-accepted: ""
+ticket-url: ""
+implemented: ""
+---
+
+This proposal is [discussed at this pull request](https://github.com/ghc-proposals/ghc-proposals/pull/597).
+
+# Allow constraint tuples and synonyms in typeclass instances
+
+## Motivation
+
+Most of Haskell projects do a lot of `stock` and `anyclass` deriving
+for repeating classes. Most of time there is some default list
+of classes to be defined.
+
+Making type synonym for that seems like the obvious thing to do,
+because jet it does not work.
+
+There are two constructions forbidden now, but required for that usecase:
+
+* Constraint synonyms
+* Constraint tuples
+
+Synonyms worked at some moment,
+but was disabled due to wrong behaviour in some cases.
+Reasoning for that is covered in Note "Instances and constraint synonyms"
+and [issue 13267](https://gitlab.haskell.org/ghc/ghc/-/issues/13267).
+
+Main usecase only affect deriving instances head,
+but another usecase is usage of tuples/synonims in context of any instance.
+It seems to be less common, but natural to cover in same proposal
+to allow tuples/synonims anywhere in instance declaration,
+leading to more consistent semantics.
```

Comment:
```suggestion
Main usecase only affect deriving instances head,
but another usecase is usage of tuples/synonyms in context of any instance.
It seems to be less common, but natural to cover in same proposal
to allow tuples/synonyms anywhere in instance declaration,
leading to more consistent semantics.
```

Reply:
Unless I'm misunderstanding, the secondary use-case is already possible in GHC today

```
> import GHC.Types
> type ShowEq :: Type -> Constraint; type ShowEq a = (Show a, Eq a)
> instance ShowEq a => Show (a -> b) where { show _ = "example" }
```